myia/compile/backends/relay.py

```diff
@@ -0,0 +1,370 @@
+"""Transforms a graph into lower-level code."""
+
+import tvm
+from tvm import relay
+import numpy as np
+
+from . import Backend
+
+from ...abstract import AbstractArray, AbstractTuple, AbstractScalar, \
+    AbstractFunction, VirtualFunction, GraphFunction, TypedPrimitive, \
+    PartialApplication, SHAPE, build_type
+from ...ir import manage
+from ...graph_utils import toposort
+from ...prim import Primitive, ops as P
+from ...dtype import type_to_np_dtype, ismyiatype, Bool, Tuple
+from ...utils import overload
+
+from ..transform import set_types
+from .relay_helpers import optimize, build_module
+
+
+@overload(bootstrap=True)
+def to_relay_type(self, a: AbstractScalar):
+    """Convert a myia abstract to a Relay type."""
+    tp = build_type(a)
+    if ismyiatype(tp, Bool):
+        return relay.ty.TensorType((), 'bool')
+    else:
+        return relay.ty.TensorType((), type_to_np_dtype(tp))
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: AbstractTuple):
+    return relay.ty.TupleType([self(e) for e in a.elements])
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: AbstractArray):
+    tp = build_type(a.element)
+    return relay.ty.TensorType(a.values[SHAPE], type_to_np_dtype(tp))
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: AbstractFunction):
+    return self(a.get_unique())
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: (VirtualFunction, TypedPrimitive)):
+    return relay.ty.FuncType([self(aa) for aa in a.args],
+                             self(a.output))
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: PartialApplication):
+    tp = self(a.fn)
+    return relay.ty.FuncType(tp.arg_types[len(a.args):], tp.ret_type)
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: GraphFunction):
+    return self(a.graph.abstract)
+
+
+@overload  # noqa: F811
+def to_relay_type(self, a: object):  # pragma: no cover
+    raise ValueError("Unknown type:", build_type(a))
+
+
+def ashape(node):
+    """Make sure shape isn't None, that makes relay crash later."""
+    shp = node.shape
+    if shp is None:  # pragma: no cover
+        raise RuntimeError("You found a way to trigger this, please report it")
+        shp = ()
```

Comment:
Is `shp = ()` a legitimate thing to do if this is triggered? If shape is None, this is not an array.

Reply:
Is shape is None, then it is a scalar.  But scalars are defined as 0-dim arrays for Relay.

We may want to do something differently eventually once there is better support for actual scalars.